### PR TITLE
OCM-7184 | feat: Moving create machinepool and nodepool functions from cmd -> pkg

### PR DESCRIPTION
**Pull request by den-rgb**
Moving create nodepool + machinepool  functions from `cmd` directory to the `pkg/machinepool` directory, updating tests to increase coverage to +70%
[OCM-7184 ](https://issues.redhat.com/browse/OCM-7184)

### CODE REVIEW

**Comment by robpblake** [comment]
Holding for cut of `1.2.41` release.

File: pkg/machinepool/machinepool.go [hunk 1]
```diff
@@ -19,11 +34,43 @@ import (
...
 	DescribeMachinePool(r *rosa.Runtime, cluster *cmv1.Cluster, clusterKey string, machinePoolId string) error
 	ListMachinePools(r *rosa.Runtime, clusterKey string, cluster *cmv1.Cluster) error
 	DeleteMachinePool(r *rosa.Runtime, machinePoolId string, clusterKey string, cluster *cmv1.Cluster) error
+	AddMachinePool(cmd *cobra.Command, clusterKey string, cluster *cmv1.Cluster, r *rosa.Runtime,
+		args *MachinePoolArgs) error
+	AddNodePool(cmd *cobra.Command, clusterKey string, cluster *cmv1.Cluster, r *rosa.Runtime,
+		args *MachinePoolArgs) error
+}
+
+type MachinePoolArgs struct {
```
Comment (ciaranRoche) [inline pkg/machinepool/machinepool.go:52]: NP : move this to a different file. This file is going to get very large and we can easily move this to its own file

File: pkg/machinepool/machinepool.go [hunk 2]
```diff
@@ -35,6 +82,1061 @@ func NewMachinePoolService() MachinePoolService {
...
+		r.Reporter.Infof("Machine pool '%s' created successfully on cluster '%s'", name, clusterKey)
+		r.Reporter.Infof("To view the machine pool details, run 'rosa describe machinepool --cluster %s --machinepool %s'",
+			clusterKey, name)
+		r.Reporter.Infof("To view all machine pools, run 'rosa list machinepools --cluster %s'", clusterKey)
+	}
+
+	return nil
+}
+
+func Split(r rune) bool {
```
Comment (ciaranRoche) [inline pkg/machinepool/machinepool.go:573]: Pull this out into the helper

File: pkg/machinepool/machinepool.go [hunk 3]
```diff
@@ -35,6 +82,1061 @@ func NewMachinePoolService() MachinePoolService {
...
+	price, err := strconv.ParseFloat(spotMaxPrice, commonUtils.MaxByteSize)
+	if err != nil {
+		return fmt.Errorf("Expected a numeric value for spot max price")
+	}
+
+	if price <= 0 {
+		return fmt.Errorf("Spot max price must be positive")
+	}
+	return nil
+}
```
Comment (ciaranRoche) [inline pkg/machinepool/machinepool.go:645]: NP : might be worth breaking these out into their own file, try bundle the machine pool validations together

File: pkg/machinepool/machinepool_test.go [hunk 4]
```diff
@@ -225,6 +227,49 @@ var _ = Describe("Machinepool and nodepool", func() {
 				),
 			)
 		})
+		It("should return an error if both `subnet` and `availability-zone` flags are set", func() {
```
Comment (ciaranRoche) [inline pkg/machinepool/machinepool_test.go:230]: This context is getting too large, lets create another one for just testing creating node pools

```suggestion
		Context("Create NodePool", func() {
			When("both `subnet` and `availability-zone` flags are set", func() {
				It("should return an error", func() {
					....
				})
			})
		})
```
Lets also make use of `When` and `It`, to help with readability

File: pkg/machinepool/machinepool_test.go [hunk 5]
```diff
@@ -225,6 +227,49 @@ var _ = Describe("Machinepool and nodepool", func() {
...
+			cmd := &cobra.Command{}
+			cmd.Flags().Bool("availability-zone", true, "")
+			cmd.Flags().Bool("subnet", true, "")
+
+			clusterKey := "test-cluster-key"
+			clusterBuilder := cmv1.NewCluster().ID("test").State(cmv1.ClusterStateReady)
+			cluster, err := clusterBuilder.Build()
```
Comment (ciaranRoche) [inline pkg/machinepool/machinepool_test.go:237]: Can we update this to use the MockCluster func :pray: 
```suggestion
			cluster := MockCluster(func(c *cmv1.ClusterBuilder) {
				c.State(cmv1.ClusterStateReady)
				b := cmv1.HypershiftBuilder{}
				b.Enabled(true)
				c.Hypershift(&b)
			})
```

File: pkg/machinepool/machinepool_test.go [hunk 6]
```diff
@@ -225,6 +227,49 @@ var _ = Describe("Machinepool and nodepool", func() {
...
+
+			machinePool := &machinePool{}
+			err = machinePool.AddNodePool(cmd, clusterKey, cluster, r, args)
+
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(Equal("Setting both `subnet` and " +
+				"`availability-zone` flag is not supported. Please select `subnet` " +
+				"or `availability-zone` to create a single availability zone machine pool"))
+		})
+		It("should fail name validation", func() {
```
Comment (ciaranRoche) [inline pkg/machinepool/machinepool_test.go:253]: Same as above, use `When` and `It` to add more details as to why the validation is failing

File: pkg/machinepool/machinepool_test.go [hunk 7]
```diff
@@ -225,6 +227,49 @@ var _ = Describe("Machinepool and nodepool", func() {
...
+			cmd := &cobra.Command{}
+			r := &rosa.Runtime{}
+			args := &MachinePoolArgs{}
+			machinePool := &machinePool{}
+
+			clusterKey := "test-cluster-key"
+			clusterBuilder := cmv1.NewCluster().ID("test").State(cmv1.ClusterStateReady)
```
Comment (ciaranRoche) [inline pkg/machinepool/machinepool_test.go:260]: Same as above replace with MockCluster func

File: pkg/machinepool/machinepool_test.go [hunk 8]
```diff
@@ -307,5 +352,179 @@ var _ = Describe("Machinepool and nodepool", func() {
 			Expect(MachinePoolKeyRE.MatchString("m123123123123123123123123123")).To(BeTrue())
 			Expect(MachinePoolKeyRE.MatchString("m#123")).To(BeFalse())
 		})
+		It("Tests getMachinePoolAvailabilityZones", func() {
```
Comment (ciaranRoche) [inline pkg/machinepool/machinepool_test.go:355]: This is in the context of `editMachinePoolAutoscaling` is this the correct place? 

IMO create a new `Context` for testing `getMachinePoolAvailabilityZones`

File: pkg/machinepool/machinepool_test.go [hunk 9]
```diff
@@ -307,5 +352,179 @@ var _ = Describe("Machinepool and nodepool", func() {
...
+			cluster    *cmv1.Cluster
+			err        error
+		)
+
+		JustBeforeEach(func() {
+			cmd = &cobra.Command{}
+			r = &rosa.Runtime{}
+			args = &MachinePoolArgs{}
+			clusterKey = "test-cluster-key"
+			clusterBuilder := cmv1.NewCluster().ID("test").State(cmv1.ClusterStateReady)
```
Comment (ciaranRoche) [inline pkg/machinepool/machinepool_test.go:404]: Lets replace with our MockCluster func :pray:

File: pkg/machinepool/machinepool_test.go (hunk 4 above)
Comment (den-rgb) [reply pkg/machinepool/machinepool_test.go:230]: I'll remove these tests for now and add them as part of a seperate pr with the new service changes

File: pkg/machinepool/helper_test.go [hunk 10]
```diff
@@ -120,9 +126,123 @@ var _ = Describe("Machine pool helper", func() {
...
+		r = &rosa.Runtime{AWSClient: mockClient}
+		cmd = &cobra.Command{}
+		az = "us-east-1a"
+		subnetId1 = "subnet-123"
+		subnetId2 = "subnet-456"
+		args = &MachinePoolArgs{}
+	})
+
+	Context("When testing getSubnetFromAvailabilityZone", func() {
+		It("Tests getSubnetFromAvailabilityZone with no availability zone set and one subnet returned", func() {
```
Comment (ciaranRoche) [inline pkg/machinepool/helper_test.go:217]: Can we flesh these test descriptions out a little, its kinda hard to understand what we are testing for
```suggestion
	       When("no availability zone is set", func() {
		          It("is should return available subnets", func() {
```

File: pkg/machinepool/helper_test.go [hunk 11]
```diff
@@ -120,9 +126,123 @@ var _ = Describe("Machine pool helper", func() {
...
+				b.Enabled(true)
+				c.Hypershift(&b).
+					Nodes(cmv1.NewClusterNodes().AvailabilityZones("us-east-1a")).AWS(cmv1.NewAWS().SubnetIDs(subnetId1, subnetId2))
+			})
+			subnet, err := getSubnetFromAvailabilityZone(cmd, r, false, cluster, *args)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(subnet).To(Equal(subnetId1))
+		})
+
+		It("Tests error case for getSubnetFromAvailabilityZone with availability zone set", func() {
```
Comment (ciaranRoche) [inline pkg/machinepool/helper_test.go:232]: An example of an improvement based on the comment above
```suggestion
		What("an availability zone is set", func() {
		        It("it should return available subnets", func() {
```

File: pkg/machinepool/helper.go [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+			os.Exit(1)
+		}
+		subnet = aws.ParseOption(subnetOption)
+	}
+
+	return subnet
+}
+
+// getSubnetOptions gets one of the cluster subnets and returns a slice of formatted VPC's private subnets.
+func getSubnetOptions(r *rosa.Runtime, cluster *cmv1.Cluster) ([]string, error) {
```
Comment (ciaranRoche) [inline pkg/machinepool/helper.go:74]: NP : Can we get a UT for this please :pray:  I know its implicitly tested in `getSubnetFromUser` but lets be pedantic on this :pray:

File: pkg/machinepool/helper.go [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+	// Format subnet options
+	var subnetOptions []string
+	for _, subnet := range privateSubnets {
+		subnetOptions = append(subnetOptions, aws.SetSubnetOption(subnet))
+	}
+
+	return subnetOptions, nil
+}
+
+func getSecurityGroupsOption(r *rosa.Runtime, cmd *cobra.Command, cluster *cmv1.Cluster) ([]string, error) {
```
Comment (ciaranRoche) [inline pkg/machinepool/helper.go:90]: Needs a unit test :pray:

File: pkg/machinepool/helper.go [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+	if cluster.Hypershift().Enabled() {
+		id = cluster.ID()
+	} else {
+		id = cluster.InfraID()
+	}
+
+	return interactiveSgs.GetSecurityGroupIds(r, cmd, vpcId, interactiveSgs.MachinePoolKind, id), nil
+}
+
+func createAwsNodePoolBuilder(
```
Comment (ciaranRoche) [inline pkg/machinepool/helper.go:115]: Also missing a unit test, if you can add one for this too :pray:

File: pkg/machinepool/helper.go [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+	}
+
+	if len(awsTags) > 0 {
+		awsNpBuilder.Tags(awsTags)
+	}
+
+	return awsNpBuilder
+}
+
+func getVpcIdFromSubnet(subnet ec2types.Subnet) (string, error) {
```
Comment (ciaranRoche) [inline pkg/machinepool/helper.go:133]: Missing unit test for this function also

**Review by ciaranRoche — COMMENTED** [review]
Looks good, can you add unit tests for the helpers, and we can merge, and follow up with the machine pool tests in a different MR. 

Also left some comments on structuring the tests

**Review by ciaranRoche — APPROVED** [review]
Approving this one, lets get it through QE regression and follow up with the unit tests.

**Pull request merged**